packages/vite/src/node/plugins/define.ts

```diff
@@ -30,31 +30,54 @@ export function definePlugin(config: ResolvedConfig): Plugin {
     })
   }
 
-  const replacements: Record<string, string | undefined> = {
-    'process.env.NODE_ENV': JSON.stringify(process.env.NODE_ENV || config.mode),
-    'global.process.env.NODE_ENV': JSON.stringify(
-      process.env.NODE_ENV || config.mode
-    ),
-    'globalThis.process.env.NODE_ENV': JSON.stringify(
-      process.env.NODE_ENV || config.mode
-    ),
-    ...userDefine,
-    ...importMetaKeys,
-    'process.env.': `({}).`,
-    'global.process.env.': `({}).`,
-    'globalThis.process.env.': `({}).`
+  function generatePattern(
+    ssr: boolean
+  ): [Record<string, string | undefined>, RegExp] {
+    const processEnv = ssr
+      ? {
+          // account for non-node environments like v8
+          'process.env.': `(typeof process === 'undefined' ? {} : process.env).`,
+          'global.process.env.': `(typeof global.process === 'undefined' ? {} : global.process.env).`,
+          'globalThis.process.env.': `(typeof globalThis.process === 'undefined' ? {} : globalThis.process.env).`
+        }
```

Comment:
I'm extremely unhappy with this runtime check, but as @frandiox points out, potential target environments like Cloudflare Workers and Deno Deploy do not have `process` as it is a node-ism.

Information about the target environment would need to be available at build time in order to remove the runtime check.

Reply:
@GrygrFlzr I believe there's a build variable in Vite, `ssr.target`, which you might be able to use for this. It can have `webworker` value. Therefore, the check would happen at build-time instead of run-time like in this code 🤔 

Not sure if it's easy to access that variable though.